src/iop/mod.rs

```diff
@@ -327,53 +320,60 @@ impl From<SynthesisError> for Error {
+/// `L_H(X,Y) = 1/|H| * (Y*V_H(X) - X*V_H(Y)) / (X-Y)`
+/// where `V_H()` is the vanishing polynomial over `H`.
+pub trait LagrangeKernel<F: PrimeField> {
+    /// Succinct evaluation of `L(x,y)`.
+    fn eval_lagrange_kernel(&self, x: F, y: F) -> F;
 
-    /// Evaluate the magic polynomial over `self`
-    fn batch_eval_unnormalized_bivariate_lagrange_poly_with_same_inputs(&self) -> Vec<F>;
+    /// Domain evaluation of `L(X,y)`, for `y` not being from `H`.
+    fn domain_eval_lagrange_kernel(&self, y: F) -> Result<Vec<F>, Error>;
 }
 
-impl<F: PrimeField> UnnormalizedBivariateLagrangePoly<F> for Box<dyn EvaluationDomain<F>> {
-    fn eval_unnormalized_bivariate_lagrange_poly(&self, x: F, y: F) -> F {
+impl<F: PrimeField> LagrangeKernel<F> for Box<dyn EvaluationDomain<F>> {
+    // L_H(x,y) = (y*V_H(x) - x*V_H(y)) / (x-y) for x!=y,
+    // L_H(x,x) = (1 + (n - 1) * x^n) / n.
+    // In particular, if both x and y belong to H,
+    // L_H(x,y) = 0 for x!=y
+    // L_H(x,x) = 1
+    fn eval_lagrange_kernel(&self, x: F, y: F) -> F {
+        let n = self.size_as_field_element();
+        let n_inv = self.size_inv();
         if x != y {
-            (self.evaluate_vanishing_polynomial(x) - self.evaluate_vanishing_polynomial(y))
+            n_inv
+                * (y * self.evaluate_vanishing_polynomial(x)
+                    - x * self.evaluate_vanishing_polynomial(y))
                 / (x - y)
         } else {
-            self.size_as_field_element() * x.pow(&[(self.size() - 1) as u64])
+            n_inv * (F::one() + (n - F::one()) * x.pow(&[(self.size()) as u64]))
         }
     }
 
-    fn batch_eval_unnormalized_bivariate_lagrange_poly_with_diff_inputs(&self, x: F) -> Vec<F> {
-        let vanish_x = self.evaluate_vanishing_polynomial(x);
-        let mut inverses: Vec<F> = self.elements().map(|y| x - y).collect();
-        algebra::fields::batch_inversion(&mut inverses);
-
-        inverses
-            .par_iter_mut()
-            .for_each(|denominator| *denominator *= vanish_x);
-        inverses
-    }
-
-    fn batch_eval_unnormalized_bivariate_lagrange_poly_with_same_inputs(&self) -> Vec<F> {
-        let mut elems: Vec<F> = self
-            .elements()
-            .map(|e| e * self.size_as_field_element())
-            .collect();
-        elems[1..].reverse();
-        elems
+    // Compute L_H(X,y) = -(y^n - 1) / n * X / (X-y) using batch inversion and index reversion.
+    // Costs essentially a vector product and a batch inversion.
+    fn domain_eval_lagrange_kernel(&self, y: F) -> Result<Vec<F>, Error> {
+        let v_at_y = self.evaluate_vanishing_polynomial(y);
+        if v_at_y.is_zero() {
+            Err(Error::Other(
+                "argument should not belong to the domain".to_owned(),
+            ))
+        } else {
+            let c1 = -self.size_as_field_element() / v_at_y;
+            let c2 = -c1 * y;
+            let mut inverses: Vec<_> = self.elements().map(|x| c1 + c2 * x).collect();
```

Comment:
Since `c1` is a constant value, adding `c1` to all the elements of the vector and then rearranging its elements is equivalent to first rearrange the elements of the vector and then adding `c1` to all its elements